polynote-frontend/polynote/ui/component/nb_config.ts

```diff
@@ -0,0 +1,382 @@
+                    para([], ['You can also specify pip packages, e.g. ', span(['pre'], ['requests']), ', or with a version like ', span(['pre'], ['urllib3==1.25.3'])]),
+                    this.configHandler.dependencyContainer
+                ]),
+                div(['notebook-resolvers', 'notebook-config-section'], [
+                    h3([], ['Resolvers']),
+                    para([], ['Specify any custom Ivy, Maven, or Pip repositories here.']),
+                    this.configHandler.resolverContainer
+                ]),
+                div(['notebook-exclusions', 'notebook-config-section'], [
+                    h3([], ['Exclusions']),
+                    para([], ['[Scala only]: Specify organization:module coordinates for your exclusions, i.e. ', span(['pre'], ['org.myorg:package-name_2.11'])]),
+                    this.configHandler.exclusionContainer
+                ]),
+                div(['notebook-spark-config', 'notebook-config-section'], [
+                    h3([], ['Spark Config']),
+                    para([], ['Set Spark configuration for this notebook here. Please note that it is possible that your environment may override some of these settings at runtime :(']),
+                    this.configHandler.sparkConfigContainer
+                ]),
+                div(['controls'], [
+                    button(['save'], {}, ['Save & Restart']).click(evt => {
+                        const conf = this.configHandler.toConfig();
+                        this.lastConfig = conf;
+                        this.el.classList.remove("open");
+                        this.dispatchEvent(new UIEvent('UpdatedConfig', {config: conf}));
+                    }),
+                    button(['cancel'], {}, ['Cancel']).click(evt => {
+                        if (this.lastConfig) {
+                            this.setConfig(this.lastConfig);
+                        }
+                        this.el.classList.remove("open");
+                    })
+                ])
+            ])
+        ];
+
+        while (this.el.childNodes.length > 0) {
+            this.el.removeChild(this.el.childNodes[0]);
+        }
+
+        children.forEach(el => this.el.appendChild(el));
+    }
+}
+
+/**
+ * Handle changes to the notebook config, propagating them to the config UI.
+ */
+interface ConfigEl {
+    row: TagElement<"div"> | null
+}
+interface Dep extends ConfigEl {
+    data: {
+        lang: string,
+        dep: string
+    }
+}
+
+interface Excl extends ConfigEl {
+    data: {
+        exclusion: string
+    }
+}
+
+interface Res extends ConfigEl {
+    data: {
+        type: string,
+        url: string,
+        pattern?: string,
+        metadata?: string
+    }
+}
+
+interface SparkConf extends ConfigEl {
+    data: {
+        key: string,
+        val: string
+    }
+}
+class NotebookConfigHandler extends UIEventTarget {
```

Comment:
This is actually a more significant refactor than just a straight port like most of the other changes in this PR. 

Essentially, `NotebookConfigHandler` backs the config changes rather than going through the DOM (it does update the DOM to reflect changes of course).